pkg/network/setup/netstat.go

```diff
@@ -114,44 +114,58 @@ func (c *NetStat) UpdateStatus(vmi *v1.VirtualMachineInstance, domain *api.Domai
 
 	primaryNetwork := netvmispec.LookupPodNetwork(vmi.Spec.Networks)
-	interfacesStatus = restorePodIfaceNameForPrimaryIface(interfacesStatus, primaryNetwork, vmi.Status.Interfaces)
+	if primaryNetwork != nil {
+		interfacesStatus = restorePrimaryIfaceStatus(interfacesStatus, vmi.Status.Interfaces, primaryNetwork.Name)
+	}
 	primaryInterfaceStatus, interfacesStatus := netvmispec.PopInterfaceByNetwork(interfacesStatus, primaryNetwork)
 	if primaryInterfaceStatus != nil {
 		interfacesStatus = append([]v1.VirtualMachineInstanceNetworkInterface{*primaryInterfaceStatus}, interfacesStatus...)
 	}
 
 	interfacesStatus = ifacesStatusFromMultus(interfacesStatus, multusStatusNetworksByName, vmiInterfacesSpecByName)
 
+	interfacesStatus = restorePodIfaceNames(interfacesStatus, vmi.Status.Interfaces)
 	vmi.Status.Interfaces = interfacesStatus
 
 	c.removeAbsentIfacesFromVolatileCache(vmi)
 
 	return nil
 }
 
-func restorePodIfaceNameForPrimaryIface(
+// restorePrimaryIfaceStatus restores the primary interface's status in case it was previously reported
+func restorePrimaryIfaceStatus(
 	interfacesStatus []v1.VirtualMachineInstanceNetworkInterface,
-	primaryNetwork *v1.Network,
 	prevIfaceStatuses []v1.VirtualMachineInstanceNetworkInterface,
+	primaryNetworkName string,
 ) []v1.VirtualMachineInstanceNetworkInterface {
-	if primaryNetwork == nil {
+	if primaryIfaceStatus := netvmispec.LookupInterfaceStatusByName(interfacesStatus, primaryNetworkName); primaryIfaceStatus != nil {
 		return interfacesStatus
 	}
 
-	prevPrimaryIfaceStatus := netvmispec.LookupInterfaceStatusByName(prevIfaceStatuses, primaryNetwork.Name)
+	prevPrimaryIfaceStatus := netvmispec.LookupInterfaceStatusByName(prevIfaceStatuses, primaryNetworkName)
 	if prevPrimaryIfaceStatus == nil {
 		return interfacesStatus
 	}
 
-	primaryIfaceStatus := netvmispec.LookupInterfaceStatusByName(interfacesStatus, primaryNetwork.Name)
-	if primaryIfaceStatus == nil {
-		return slices.Insert(interfacesStatus, 0, v1.VirtualMachineInstanceNetworkInterface{
-			Name:             prevPrimaryIfaceStatus.Name,
-			PodInterfaceName: prevPrimaryIfaceStatus.PodInterfaceName,
-		})
+	return slices.Insert(interfacesStatus, 0, v1.VirtualMachineInstanceNetworkInterface{
+		Name: prevPrimaryIfaceStatus.Name,
+	})
+}
+
+// restorePodIfaceNames restores the PodInterfaceName based on the VMI controller's last report
+func restorePodIfaceNames(
+	interfacesStatus []v1.VirtualMachineInstanceNetworkInterface,
+	prevIfaceStatuses []v1.VirtualMachineInstanceNetworkInterface,
+) []v1.VirtualMachineInstanceNetworkInterface {
+	prevIfaceStatusesByName := netvmispec.IndexInterfaceStatusByName(prevIfaceStatuses, netvmispec.HasIfaceStatusOriginatedFromSpec)
+
+	for i, ifaceStatus := range interfacesStatus {
+		if !netvmispec.HasIfaceStatusOriginatedFromSpec(ifaceStatus) {
+			continue
+		}
+		interfacesStatus[i].PodInterfaceName = prevIfaceStatusesByName[ifaceStatus.Name].PodInterfaceName
```

Comment:
The last line here is correct because we exit one step before.
I find it very hard to compute this.. it is indirect.

A direct approach would be to perform this operation only if it is possible.
In fact, this looks much clearer:
```go
if ifaceStatus.Name != nil {
    interfacesStatus[i].PodInterfaceName = prevIfaceStatusesByName[ifaceStatus.Name].PodInterfaceName
}
```

You can explain in the function doc how to identify when the information is based on the spec.

Reply:
Done.